service/worker/deployment/deployment_workflow.go

```diff
@@ -25,91 +25,128 @@
-	updateBuildIDSignalChannel := workflow.GetSignalChannel(d.ctx, UpdateDeploymentBuildIDSignalName)
-	forceCANSignalChannel := workflow.GetSignalChannel(d.ctx, ForceCANSignalName)
+	updateBuildIDSignalChannel := workflow.GetSignalChannel(ctx, UpdateDeploymentBuildIDSignalName)
+	forceCANSignalChannel := workflow.GetSignalChannel(ctx, ForceCANSignalName)
 	forceCAN := false
 
-	selector := workflow.NewSelector(d.ctx)
-	selector.AddReceive(updateDeploymentSignalChannel, func(c workflow.ReceiveChannel, more bool) {
-		// Process Signal
-		forceCAN = true
-	})
+	selector := workflow.NewSelector(ctx)
 	selector.AddReceive(updateBuildIDSignalChannel, func(c workflow.ReceiveChannel, more bool) {
 		// Process Signal
 	})
 	selector.AddReceive(forceCANSignalChannel, func(c workflow.ReceiveChannel, more bool) {
 		// Process Signal
+		forceCAN = true
 	})
 
-	// async draining before CAN
-	for (!workflow.GetInfo(d.ctx).GetContinueAsNewSuggested() && !forceCAN) || selector.HasPending() {
-		selector.Select(d.ctx)
+	for (!workflow.GetInfo(ctx).GetContinueAsNewSuggested() && !forceCAN) || selector.HasPending() {
+		selector.Select(ctx)
+	}
+
+	// Done processing signals before CAN
+	a.SignalsCompleted = true
+}
+
+func (d *DeploymentWorkflowRunner) run() error {
+	var a AwaitSignals
+	var pendingUpdates int
+
+	// Set up Query Handlers here:
+	err := workflow.SetQueryHandler(d.ctx, "deploymentTaskQueues", func(input []byte) (map[string]*deployspb.DeploymentLocalState_TaskQueueFamilyInfo, error) {
+		return d.DeploymentLocalState.TaskQueueFamilies, nil
+	})
+	if err != nil {
+		d.logger.Error("Failed while setting up query handler")
+		return err
+	}
+
+	// Listen to signals in a different go-routine to make business logic clearer
+	workflow.Go(d.ctx, a.ListenToSignals)
+
+	// Setting an update handler for updating deployment task-queues
+	if err := workflow.SetUpdateHandler(
+		d.ctx,
+		RegisterWorkerInDeployment,
+		func(ctx workflow.Context, updateInput *deployspb.RegisterWorkerInDeploymentArgs) error {
+			pendingUpdates++
+			defer func() {
+				pendingUpdates--
+			}()
+
+			if d.DeploymentLocalState.TaskQueueFamilies == nil {
+				d.DeploymentLocalState.TaskQueueFamilies = make(map[string]*deployspb.DeploymentLocalState_TaskQueueFamilyInfo)
+			}
+			if d.DeploymentLocalState.TaskQueueFamilies[updateInput.TaskQueueName] == nil {
+				d.DeploymentLocalState.TaskQueueFamilies[updateInput.TaskQueueName] = &deployspb.DeploymentLocalState_TaskQueueFamilyInfo{}
+			}
+
+			// Add the task queue to the local state
+			newTaskQueueWorkerInfo := &deployspb.DeploymentLocalState_TaskQueueFamilyInfo_TaskQueueInfo{
+				TaskQueueType:   updateInput.TaskQueueType,
+				FirstPollerTime: updateInput.FirstPollerTime,
+			}
+			d.DeploymentLocalState.TaskQueueFamilies[updateInput.TaskQueueName].TaskQueues =
+				append(d.DeploymentLocalState.TaskQueueFamilies[updateInput.TaskQueueName].TaskQueues, newTaskQueueWorkerInfo)
```

Comment:
what if there already is an entry in this list for this task queue type?

Reply:
this question deals with making the update idempotent in nature right (we want to prevent multiple of these duplicate entries since technically, we could get multiple repeated update-requests)

I have thought about this problem but not addressed it right now - shall be doing so in a follow-up PR - wanted this to be a basic feel about how the workflow would look like (I should probably add a TODO here for a potential reviewer)

Reply:
This will be part of the follow-up PR which will be all about the updateHandler - starting the activity from within, having locks in the updateHandler, and making update to the local state idempotent